### PR TITLE
Introduce FHIR version-agnostic FHIR JSON to Def Integration Testing Framework With two R4 and one DSTU3 POCs

### PR DESCRIPTION
**Pull request by lukedegruchy**
# Merge Request: Version-Agnostic MeasureDef Framework with Integrated Test Infrastructure

## Summary

This merge request introduces a version-agnostic internal model (`MeasureDef` family of classes) for capturing measure evaluation state, and integrates assertion capabilities directly into the existing R4 and DSTU3 Measure/MultiMeasure test DSLs. The changes enable tests to assert on both pre-scoring internal state (MeasureDef) and post-scoring FHIR resources (MeasureReport) in a unified fluent API, while eliminating the need for external callbacks or deep object copying. The implementation consists of two phases: (1) establishing the foundational MeasureDef framework with population basis propagation, and (2) enhancing test infrastructure to support dual assertions without parallel test systems.

Key architectural changes:

1. **Version-Agnostic MeasureDef Framework**: Leveraged `MeasureDef`, `GroupDef`, `PopulationDef`, `StratifierDef`, `StratumDef` classes as internal models that capture measure evaluation state independently of FHIR version, enabling consistent processing logic across DSTU3, R4, and R5.

2. **Population Basis Propagation**: Modified `PopulationDef` to own its `populationBasis` field (previously accessed through `GroupDef`), eliminating parameter passing in `getCount()` method and simplifying scoring logic throughout `MeasureDefScorer`.

3. **Test Infrastructure Integration**: Added `@VisibleForTesting` methods (`evaluateMeasureCaptureDefs()`, `evaluateWithDefs()`) to processors and services that return record types pairing `MeasureDef` with `MeasureReport`, enabling test frameworks to access both pre-scoring and post-scoring state.

4. **Enhanced Test DSL**: Extended R4/DSTU3 Measure and MultiMeasure test DSLs with `def()` and `defs()` methods for asserting on MeasureDef hierarchy, complementing existing `report()` methods for MeasureReport assertions, with full `up()` navigation support for fluent test chains.

5. **Extracted Test Framework Classes**: Refactored inner Selected* classes from Measure.java into dedicated packages (`r4/selected/report/` and `r4/selected/def/`) with clearer naming (`SelectedMeasureReport*` vs `SelectedMeasureDef*`), reducing Measure.java from 987 lines to 115 lines.

## Code Review Suggestions

- [ ] **Population Basis Ownership in PopulationDef**: Verify that `PopulationDef.getCount()` correctly uses its own `populationBasis` field instead of requiring `GroupDef` parameter. Check that boolean basis (`isBooleanBasis()`) vs resource basis counting logic is correct for all population types, especially MEASURE-OBSERVATION populations which count observations differently. Review how this interacts with continuous variable measures that have aggregate functions.

- [ ] **@VisibleForTesting Method Thread Safety**: Confirm that the assumption of synchronous, single-threaded evaluation holds across all usage contexts (R4MeasureService, R4MultiMeasureService, batch operations). The `@VisibleForTesting` methods return mutable `MeasureDef` references that are safe only if no concurrent modification occurs. Verify this is documented and enforced.

- [ ] **MeasureDef Builder Population Basis Propagation**: Review `R4MeasureDefBuilder.buildPopulationDef()` and `Dstu3MeasureDefBuilder.buildPopulationDef()` to ensure `populationBasis` is correctly resolved from measure-level and group-level basis, and properly passed to all `PopulationDef` constructors including date-of-compliance populations.

- [ ] **MeasureDefScorer Count Method Signature Change**: Audit all calls to `PopulationDef.getCount()` in `MeasureDefScorer` (proportion, ratio, continuous-variable, cohort scoring) to ensure removal of `GroupDef` parameter doesn't break scoring logic. Pay special attention to stratified scoring where both group-level and stratum-level counts are calculated.

- [ ] **Multi-Measure Bundling in R4MultiMeasureService**: Verify that `evaluateWithDefs()` correctly builds `Parameters` with bundled `MeasureReport` resources according to evaluation type (population vs subject). For population evaluation, ensure one bundle contains all measure reports. For subject evaluation, ensure one bundle per subject containing all their measure reports. Check that `MeasureDefAndR4ParametersWithMeasureReports` record correctly pairs the List<MeasureDef> with the Parameters structure.

- [ ] **Test DSL Backward Compatibility**: Confirm existing tests using old API (e.g., `.then().firstGroup()`) continue to work via delegation methods. Check that `Then` class properly delegates to `report()` for methods like `firstGroup()`, `group(String)`, `hasGroupCount()`, etc. Verify no behavioral changes in existing test assertions.

- [ ] **Selected Class Parent Type Parameters**: Review generic parent type parameters (`<P>`) in Selected* classes to ensure `up()` navigation works correctly across hierarchies. For example, `SelectedMeasureDefPopulation<SelectedMeasureDefGroup<SelectedMeasureDef<Then>>>` should allow `.up().up().up()` to return to `Then`, enabling switching between `def()` and `report()` assertions.

- [ ] **Version-Agnostic MeasureDef Usage**: Verify that `MeasureDef` classes in `common` package truly work across DSTU3, R4, and R5 without version-specific dependencies. Check that `MeasureDefBuilder` implementations (R4, DSTU3) correctly populate the shared `MeasureDef` model and that `MeasureReportBuilder` implementations (R4, DSTU3) correctly consume it for scoring.

- [ ] **Production Code Isolation**: Confirm that `@VisibleForTesting` methods are package-private and only called from test code. Verify production paths (public `evaluate()` methods) delegate to test-visible methods and extract only `MeasureReport`, never exposing `MeasureDef` in production APIs. Check that no production code depends on test-only record types.

### Setup

1. **Test Environment**: Use HAPI JPA server instance with FHIR R4 endpoint
2. **Test Measures**: Load the following measure types into repository:
   - Proportion measure with boolean basis (patient-based counting)
   - Proportion measure with resource basis (encounter-based counting)
   - Continuous variable measure with aggregate functions (avg, median, sum)
   - Ratio measure with multiple populations
   - Composite measure with multiple component measures
   - Stratified measure with 2+ stratifiers
3. **Test Data**: Load corresponding patient data, observations, encounters for each measure type
4. **Evaluation Parameters**: Prepare test cases with varying:
   - Measurement periods (past, current, future dates)
   - Report types (individual, summary, subject-list)
   - Subject IDs (single patient, multiple patients, practitioner)
   - Additional data bundles

### Test Cases

- [ ] **Boolean Basis Proportion Measure - Individual Report**: Evaluate a proportion measure with `populationBasis="boolean"` for a single patient. Verify MeasureReport populations (initial-population, denominator, numerator) count unique subjects (1 or 0). Verify MeasureReport.measureScore is correctly calculated as numerator/denominator. Cross-check with any existing measure evaluation results to ensure no regression.

- [ ] **Resource Basis Proportion Measure - Individual Report**: Evaluate a proportion measure with `populationBasis="Encounter"` for a single patient with multiple encounters. Verify MeasureReport populations count all matching encounters (not just 1 per patient). Verify scoring correctly uses resource counts. Compare with boolean-basis measure to confirm different counting behavior.

- [ ] **Continuous Variable Measure - Summary Report**: Evaluate a continuous variable measure with `aggregateMethod="median"` for a population of 10+ patients. Verify MeasureReport.group.measureScore contains the correct median value calculated across all measure-observation values. Test other aggregate methods (avg, min, max, sum, count) to ensure correct calculations.

- [ ] **Multi-Measure Evaluation - Population Report**: Use `$evaluate-measure` operation with multiple measure URLs (3+ measures) and `reportType="population"`. Verify Parameters response contains exactly one Bundle with all MeasureReports inside. Verify each MeasureReport has correct measure URL, period, and population counts. Ensure no duplicate or missing reports.

- [ ] **Multi-Measure Evaluation - Subject Report**: Use `$evaluate-measure` operation with multiple measure URLs (3+ measures) and multiple subject IDs (5+ patients). Verify Parameters response contains one Bundle per subject, each containing all measure reports for that subject. Verify total MeasureReport count equals (measures × subjects). Check bundle structure matches expected subject-based grouping.

- [ ] **Stratified Measure Evaluation**: Evaluate a measure with 2+ stratifiers (e.g., age group, gender) for a population. Verify MeasureReport.group.stratifier entries exist for each stratifier with correct stratum values. Verify each stratum has correct population counts. Ensure stratum populations sum to group-level populations.

- [ ] **Date of Compliance Population**: Evaluate a care-gap measure with `cqfm-careGapDateOfCompliance` extension. Verify MeasureReport includes the date-of-compliance population with correct expression evaluation. Check that population basis propagates correctly to this synthetic population.

- [ ] **DSTU3 Measure Evaluation**: Repeat proportion measure evaluation using DSTU3 FHIR endpoint and DSTU3 measure resources. Verify MeasureReport structure matches DSTU3 specification (different from R4). Confirm same evaluation logic produces equivalent results across versions.

- [ ] **Measure with Additional Data Bundle**: Evaluate a measure while providing `additionalData` Bundle containing supplemental resources (e.g., ValueSets, patient data not in repository). Verify evaluation uses additional data correctly (check criteria expressions reference additional resources). Ensure no errors when additional data is absent but not required.

- [ ] **Measurement Period Edge Cases**: Test measure evaluation with measurement periods that:
   - Span multiple years (e.g., 2023-01-01 to 2025-12-31)
   - Use dates outside available data range
   - Have start date equal to end date (single-day period)
   - Are missing (rely on measure's default period from CQL)

   Verify appropriate handling in each case (empty reports, default period usage, validation errors).

- [ ] **Subject-List Report Type**: Evaluate a measure with `reportType="subject-list"` for a population. Verify MeasureReport includes `evaluatedResource` extensions listing all resources evaluated for each population. Check that resource references are correct and correspond to actual evaluated data.

- [ ] **Measure with Missing Library**: Attempt to evaluate a measure where the referenced Library resource is missing from repository or has no CQL/ELM content. Verify appropriate error handling with clear error message indicating which library is missing. Ensure no null pointer exceptions or cryptic errors.

- [ ] **Concurrent Evaluation Assumption**: While the code assumes single-threaded synchronous evaluation, test whether concurrent evaluations (e.g., parallel API calls to `$evaluate-measure`) cause any issues. Check for race conditions, shared state corruption, or unexpected errors. Document any concurrency limitations discovered.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureService.java [hunk 1]
```diff
@@ -57,6 +62,65 @@ public MeasureReport evaluate(
...
+            String reportType,
+            String subjectId,
+            String lastReceivedOn,
+            Endpoint contentEndpoint,
+            Endpoint terminologyEndpoint,
+            Endpoint dataEndpoint,
+            Bundle additionalData,
+            Parameters parameters,
+            String productLine,
+            String practitioner) {
```
Comment (JPercival) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureService.java]: thought: I see... seems like this is a case of suboptimal layering, right? A MeasureDef ought be to be able to be evaluated on its own, without any reference to a FHIR measure.

Comment (lukedegruchy) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureService.java]: That's the ultimate goal:  I'd like to tackle this in a separate PR, since Justin asked me to do invasive changes to my original design, so I'd like to get this merged.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/MeasureDefAndDstu3MeasureReport.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package org.opencds.cqf.fhir.cr.measure.dstu3;
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/MeasureDefAndDstu3MeasureReport.java]: I think you figured this out judging by your next comment:  It's a part of an optimization designed to enable FHIR to Def integration testing.
